cmd/kubectl-gadget/utils/trace.go

```diff
@@ -464,61 +467,102 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
 // satisfy the conditionFunction received as parameter.
 func waitForCondition(traceID string, conditionFunction func(*gadgetv1alpha1.Trace) bool) (*gadgetv1alpha1.TraceList, error) {
 	var returnedTraces gadgetv1alpha1.TraceList
+	watchedTraces := make(map[string]gadgetv1alpha1.Trace)
+
+	traceList, err := getTraceListFromID(traceID)
+	if err != nil {
+		return nil, err
+	}
+
+	tracesNumber := len(traceList.Items)
+	tracesSatisfied := 0
 
-	tracesNumber := 0
-	watchedTracesNumber := 0
+	// Maybe some traces already satisfy conditionFunction?
+	for _, trace := range traceList.Items {
+		if conditionFunction(&trace) {
+			tracesSatisfied++
+			watchedTraces[trace.ObjectMeta.Name] = trace
+		}
+	}
+
+	// If they all satisify conditionFunction we can quit now without having to
+	// watch events.
+	if tracesSatisfied == tracesNumber {
+		for _, trace := range watchedTraces {
+			returnedTraces.Items = append(returnedTraces.Items, trace)
+		}
+
+		return &returnedTraces, nil
+	}
 
+	// Otherwise, we will need to watch events on them.
+	// For this, we will get a watcher on all the traces which share the same ID.
 	// Get a watcher on all the traces which have the same ID.
 	// Indeed, all the traces on different nodes but linked to one gadget share
 	// the same ID.
-	watcher, err := getTraceWatcher(traceID)
+	// We will also begin to monitor events since the above GET of the traces
+	// list.
+	watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
 	if err != nil {
 		return nil, err
 	}
 
 	nodeErrors := make(map[string]string)
 	nodeWarnings := make(map[string]string)
 
+	// We only watch the traces which did not satisfy conditionFunction.
+	tracesToWatch := tracesNumber - tracesSatisfied
+
 	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
 	_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
 		// This function will be executed until:
-		// 1. The number of watched traces equals the number of traces, i.e. we
-		// dealt with the traces which interest us.
+		// 1. The number of watched traces equals the number of traces to watch,
+		// i.e. we dealt with the traces which interest us.
 		// 2. Or it returns an error.
 		// 3. Or time out is fired.
 		// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
 		// with.
+		trace := event.Object.(*gadgetv1alpha1.Trace)
```

Comment:
I did this in a previous version of the code but I do not think this is necessary because the watcher only watches for this type of object.
I will nonetheless add it back, I rather have useless branch than panic.

Reply:
What you said is almost right except for the `watch.Error` that would cause a panic.